language/move-vm/runtime/src/interpreter.rs

```diff
@@ -869,6 +1033,504 @@ impl Frame {
+                    .operand_stack
+                    .pop_ty()?
+                    .check_eq(&Type::Address)?;
+                let ty = resolver.instantiate_generic_type(*idx, ty_args)?;
+                interpreter
+                    .operand_stack
+                    .push_ty(Type::MutableReference(Box::new(ty)))?;
+            }
+            Bytecode::ImmBorrowGlobalGeneric(idx) => {
+                interpreter
+                    .operand_stack
+                    .pop_ty()?
+                    .check_eq(&Type::Address)?;
+                let ty = resolver.instantiate_generic_type(*idx, ty_args)?;
+                interpreter
+                    .operand_stack
+                    .push_ty(Type::Reference(Box::new(ty)))?;
+            }
+            Bytecode::Exists(_) | Bytecode::ExistsGeneric(_) => {
+                interpreter
+                    .operand_stack
+                    .pop_ty()?
+                    .check_eq(&Type::Address)?;
+                interpreter.operand_stack.push_ty(Type::Bool)?;
+            }
+            Bytecode::MoveTo(idx) => {
+                let ty = interpreter.operand_stack.pop_ty()?;
+                interpreter
+                    .operand_stack
+                    .pop_ty()?
+                    .check_eq(&Type::Reference(Box::new(Type::Signer)))?;
+                ty.check_eq(&resolver.get_struct_type(*idx))?;
+                check_ability(resolver.loader().abilities(&ty)?.has_key())?;
+            }
+            Bytecode::MoveToGeneric(idx) => {
+                let ty = interpreter.operand_stack.pop_ty()?;
+                interpreter
+                    .operand_stack
+                    .pop_ty()?
+                    .check_eq(&Type::Reference(Box::new(Type::Signer)))?;
+                ty.check_eq(&resolver.instantiate_generic_type(*idx, ty_args)?)?;
+                check_ability(resolver.loader().abilities(&ty)?.has_key())?;
+            }
+            Bytecode::MoveFrom(idx) => {
+                interpreter
+                    .operand_stack
+                    .pop_ty()?
+                    .check_eq(&Type::Address)?;
+                let ty = resolver.get_struct_type(*idx);
+                interpreter.operand_stack.push_ty(ty)?;
+            }
+            Bytecode::MoveFromGeneric(idx) => {
+                interpreter
+                    .operand_stack
+                    .pop_ty()?
+                    .check_eq(&Type::Address)?;
+                let ty = resolver.instantiate_generic_type(*idx, ty_args)?;
+                interpreter.operand_stack.push_ty(ty)?;
+            }
+            Bytecode::FreezeRef => {
+                match interpreter.operand_stack.pop_ty()? {
+                    Type::MutableReference(ty) => {
+                        interpreter.operand_stack.push_ty(Type::Reference(ty))?
+                    }
+                    Type::Reference(ty) => {
+                        interpreter.operand_stack.push_ty(Type::Reference(ty))?
```

Comment:
I think we want this to be an error. The verifier will not allow you to apply `FreezeRef` to an immutable reference.

Reply:
There's a very interesting observation here: it seems that for vector instructions we allow implicit reference casting (mut reference can be used as immut reference). e.g: VecLen